orderbook/src/bad_token.rs

```diff
@@ -0,0 +1,577 @@
+/// simulating several transfers of a token. To find an initial address to transfer from we use
+/// the amm pair providers.
+/// Tokens are bad if:
+/// - we cannot find an amm pool of the token to one of the base tokens
+/// - transfer into the settlement contract or back out fails
+/// - a transfer loses total balance
+pub struct BadTokenDetector {
+    pub web3: Web3,
+    pub pools: Vec<Box<dyn AmmPairProvider>>,
+    pub base_tokens: HashSet<H160>,
+    pub settlement_contract: H160,
+}
+
+impl BadTokenDetector {
+    pub async fn detect(&self, token: H160) -> Result<TokenInfo> {
+        let instance = ERC20::at(&self.web3, token);
+        let decimals = match instance.decimals().call().await {
+            Ok(decimals) => decimals,
+            Err(err) => {
+                return match EthcontractError::classify(&err) {
+                    EthcontractError::Node => Err(err.into()),
+                    EthcontractError::Contract => Ok(TokenInfo::Bad {
+                        reason: "failed to get decimals".to_string(),
+                    }),
+                }
+            }
+        };
+        let amount = match U256::from(10).checked_pow(decimals.into()) {
+            Some(amount) => {
+                tracing::debug!("testing token {:?} with amount {}", token, amount);
+                amount
+            }
+            None => {
+                return Ok(TokenInfo::Bad {
+                    reason: "decimals too large".to_string(),
+                });
+            }
+        };
+
+        let take_from = match self.find_address_owning_token(token).await? {
+            Some((address, balance)) if balance > amount => {
+                tracing::debug!("testing token {:?} with pool {:?}", token, address);
+                address
+            }
+            _ => {
+                return Ok(TokenInfo::Bad {
+                    reason: "no amm pool with base tokens has enough balance".to_string(),
+                });
+            }
+        };
+
+        // We transfer one unit (base on decimals) of the token from the amm pool into the
+        // settlement contract and then to an arbitrary address.
+        // Note that gas use can depend on the recipient because for the standard implementation
+        // sending to an address that does not have any balance yet (implicitly 0) causes an
+        // allocation.
+        let request = self.create_trace_request(token, amount, take_from);
+        let traces = trace_many::trace_many(request, &self.web3).await?;
+        Self::handle_response(&traces, amount)
+    }
+
+    // Based on amm pools find the address with the largest amount of the token.
+    // Err if communication with the node failed.
+    // Ok(None) if there is no pool or getting the balance fails.
+    // Ok(address, balance) for an address that has this amount of balance of the token.
+    async fn find_address_owning_token(&self, token: H160) -> Result<Option<(H160, U256)>> {
```

Comment:
In some way that is an implementation detail because the function could find an owner of the token in another way. But I don't mind changing the name either.